README.md
# Custom JMeter sampler for authenticated S3 requests

This is a custom java sampler class that can be used to benchmark any S3 compatible system.  There is a class for GET/PUT and a class for listing.

[![GitHub tag (latest SemVer)](https://img.shields.io/github/v/tag/wheresalice/jmeter_s3_custom_sampler?sort=semver&label=Release)](https://github.com/WheresAlice/jmeter_s3_custom_sampler/releases)

* Maintained by: WheresAlice
* Additional development by: ansoni
* Originally written by: Alex Bordei @ Bigstep

## How to use

Build using maven (or download from Releases)

```shell script
mvn clean package
```

Extract zip into JMeter directory

```shell script
cd jmeter_home
unzip path/to/jmeter_s3_custom_sampler/target/*.zip
```

Run JMeter as usual

```shell script
./bin/jmeter
```

Note that upgrading isn't really supported at the moment, you will need to manually replace jars in the `lib` and `lib/ext` directories.

### S3Sampler

Add a new JMeter Java sampler, use the `protocol.java.org.apache.jmeter.protocol.java.sampler.S3Sampler` class.
![Alt text](img/jmeter1.png?raw=true "Select JMeter custom sampler")

Add your AWS key id, bucket, object and the rest.
![Alt text](img/jmeter2.png?raw=true "Configure JMeter sampler")

When reading data, if you are setting `local_file_path` then this file will be created with the contents of the S3 key.  If `local_file_path` is not set, then the content will be returned as Response Data, and will be deserialised if the content is avro format.

### S3ListSampler

Add a new JMeter Java sampler, use the `protocol.java.org.apache.jmeter.protocol.java.sampler.S3ListSampler` class.

Add your AWS key id, bucket, prefix and the rest.

Caveat: S3ListSampler will still return successful if an empty list is returned.


CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.4
### Added
- Allow username, password and scheme to be passed to proxy

## 0.3
### Added
- Return content as ResponseData when no `local_file_path` is set
- Attempt to deserialise avro when returning ResponseData
### Changed
- Dependencies updated to cope with plus symbols in S3 key names

## 0.2
### Added
- Support for listing objects in an S3 bucket

## 0.1.1
### Fixed
- Actually return metadata when using GET

## 0.1
### Added
- Writes to a file when using GET
### Changed
- Package name
- Imports cleaned up
- Typo corrected

src/protocol/java/org/apache/jmeter/protocol/java/sampler/S3ListSampler.java
package protocol.java.org.apache.jmeter.protocol.java.sampler;

import com.amazonaws.ClientConfiguration;
import com.amazonaws.Protocol;
import com.amazonaws.auth.BasicAWSCredentials;
import com.amazonaws.client.builder.AwsClientBuilder;
import com.amazonaws.internal.StaticCredentialsProvider;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.model.ListObjectsRequest;
import com.amazonaws.services.s3.model.ObjectListing;
import com.amazonaws.services.s3.model.S3ObjectSummary;
import org.apache.jmeter.config.Arguments;
import org.apache.jmeter.protocol.java.sampler.AbstractJavaSamplerClient;
import org.apache.jmeter.protocol.java.sampler.JavaSamplerContext;
import org.apache.jmeter.samplers.SampleResult;
import org.apache.jorphan.logging.LoggingManager;
import org.apache.log.Logger;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

public class S3ListSampler extends AbstractJavaSamplerClient implements Serializable {
    private static final long serialVersionUID = 2L;
    private static final Logger log = LoggingManager.getLoggerForClass();

    // set up default arguments for the JMeter GUI
    @Override
    public Arguments getDefaultParameters() {
        Arguments defaultParameters = new Arguments();
        defaultParameters.addArgument("bucket", "");
        defaultParameters.addArgument("path", "");
        defaultParameters.addArgument("key_id", "");
        defaultParameters.addArgument("secret_key", "");
        defaultParameters.addArgument("proxy_host", "");
        defaultParameters.addArgument("proxy_port", "");
        defaultParameters.addArgument("proxy_scheme", "");
        defaultParameters.addArgument("proxy_username", "");
        defaultParameters.addArgument("proxy_password", "");
        defaultParameters.addArgument("endpoint", "");
        defaultParameters.addArgument("region", "");
        return defaultParameters;
    }

    @Override
    public SampleResult runTest(JavaSamplerContext context) {
        String bucket = context.getParameter("bucket");
        String path = context.getParameter("path");
        String key_id = context.getParameter("key_id");
        String secret_key = context.getParameter("secret_key");
        String proxy_host = context.getParameter("proxy_host");
        String proxy_port = context.getParameter("proxy_port");
        String proxy_scheme = context.getParameter("proxy_scheme");
        String proxy_username = context.getParameter("proxy_username");
        String proxy_password = context.getParameter("proxy_password");
        String endpoint = context.getParameter("endpoint");
        String region = context.getParameter("region");

        SampleResult result = new SampleResult();

        result.sampleStart(); // start stopwatch
        try {
            ClientConfiguration config = new ClientConfiguration();
            if (proxy_host != null && !proxy_host.isEmpty()) {
                config.setProxyHost(proxy_host);
            }
            if (proxy_port != null && !proxy_port.isEmpty()) {
                config.setProxyPort(Integer.parseInt(proxy_port));
            }
            if (proxy_scheme != null && !proxy_scheme.isEmpty()) {
                if (proxy_scheme.toLowerCase().equals("http")) {
                    config.setProxyProtocol(Protocol.HTTP);
                } else if (proxy_scheme.toLowerCase().equals("https")) {
                    config.setProxyProtocol(Protocol.HTTPS);
                }
            }
            if (proxy_username != null && !proxy_username.isEmpty()) {
                config.setProxyUsername(proxy_username);
            }
            if (proxy_password != null && !proxy_password.isEmpty()) {
                config.setProxyPassword(proxy_password);
            }

            AmazonS3ClientBuilder builder = AmazonS3ClientBuilder.standard().withClientConfiguration(config).withRegion(region);
            if (key_id != null && !key_id.isEmpty()) {
                builder = builder.withCredentials(new StaticCredentialsProvider(new BasicAWSCredentials(key_id, secret_key)));
            }

            if (endpoint != null && !endpoint.isEmpty()) {
                builder = builder.withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(endpoint, region));
            }
            AmazonS3 s3Client = builder.build();

            ListObjectsRequest listObjectsRequest = new ListObjectsRequest()
                    .withBucketName(bucket)
                    .withPrefix(path);
            ObjectListing objectListing;
            List<String> summaries = new ArrayList<>();
            do {
                objectListing = s3Client.listObjects(listObjectsRequest);
                for (S3ObjectSummary objectSummary :
                    objectListing.getObjectSummaries()) {
                    summaries.add(objectSummary.getKey());
                }
                listObjectsRequest.setMarker(objectListing.getNextMarker());
            } while (objectListing.isTruncated());
            result.setResponseData(String.join(",", summaries), "UTF-8");
            result.sampleEnd(); // stop stopwatch
            result.setSuccessful(true);
            result.setResponseCodeOK(); // 200 code
            result.setResponseMessageOK();

        } catch (Exception e) {
            result.sampleEnd(); // stop stopwatch
            result.setSuccessful(false);
            result.setResponseMessage("Exception: " + e);

            // get stack trace as a String to return as document data
            java.io.StringWriter stringWriter = new java.io.StringWriter();
            e.printStackTrace(new java.io.PrintWriter(stringWriter));
            result.setResponseData(stringWriter.toString());
            result.setDataType(org.apache.jmeter.samplers.SampleResult.TEXT);
            result.setResponseCode("500");
        }

        return result;
    }
}


src/protocol/java/org/apache/jmeter/protocol/java/sampler/S3Sampler.java
/* Written By Alexandru Bordei
*  Bigstep.com
*  Distrubuted under the same license as apache jmeter itself.
* http://www.apache.org/licenses/LICENSE-2.0
*/
package protocol.java.org.apache.jmeter.protocol.java.sampler;


import com.amazonaws.ClientConfiguration;
import com.amazonaws.Protocol;
import com.amazonaws.auth.BasicAWSCredentials;
import com.amazonaws.client.builder.AwsClientBuilder;
import com.amazonaws.internal.StaticCredentialsProvider;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.model.*;
import org.apache.avro.Schema;
import org.apache.avro.file.DataFileStream;
import org.apache.avro.generic.GenericDatumReader;
import org.apache.avro.generic.GenericDatumWriter;
import org.apache.avro.io.DatumReader;
import org.apache.avro.io.DatumWriter;
import org.apache.avro.io.EncoderFactory;
import org.apache.avro.io.JsonEncoder;
import org.apache.jmeter.config.Arguments;
import org.apache.jmeter.protocol.java.sampler.AbstractJavaSamplerClient;
import org.apache.jmeter.protocol.java.sampler.JavaSamplerContext;
import org.apache.jmeter.samplers.SampleResult;
import org.apache.jorphan.logging.LoggingManager;
import org.apache.log.Logger;
import org.springframework.util.StreamUtils;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.Serializable;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;

public class S3Sampler extends AbstractJavaSamplerClient implements Serializable {
    private static final long serialVersionUID = 2L;
    private static final Logger log = LoggingManager.getLoggerForClass();

    // set up default arguments for the JMeter GUI
    @Override
    public Arguments getDefaultParameters() {
        Arguments defaultParameters = new Arguments();
        defaultParameters.addArgument("method", "GET");
        defaultParameters.addArgument("bucket", "");
        defaultParameters.addArgument("object", "");
        defaultParameters.addArgument("key_id", "");
        defaultParameters.addArgument("secret_key", "");
        defaultParameters.addArgument("local_file_path", "");
        defaultParameters.addArgument("proxy_host", "");
        defaultParameters.addArgument("proxy_port", "");
        defaultParameters.addArgument("proxy_scheme", "");
        defaultParameters.addArgument("proxy_username", "");
        defaultParameters.addArgument("proxy_password", "");

        defaultParameters.addArgument("endpoint", "");
        defaultParameters.addArgument("region", "");
        defaultParameters.addArgument("acl", "");
        return defaultParameters;
    }

    @Override
    public SampleResult runTest(JavaSamplerContext context) {
        // pull parameters
        String bucket = context.getParameter("bucket");
        String object = context.getParameter("object");
        String method = context.getParameter("method");
        String local_file_path = context.getParameter("local_file_path");
        String key_id = context.getParameter("key_id");
        String secret_key = context.getParameter("secret_key");
        String proxy_host = context.getParameter("proxy_host");
        String proxy_port = context.getParameter("proxy_port");
        String proxy_scheme = context.getParameter("proxy_scheme");
        String proxy_username = context.getParameter("proxy_username");
        String proxy_password = context.getParameter("proxy_password");
        String endpoint = context.getParameter("endpoint");
        String region = context.getParameter("region");
        String acl = context.getParameter("acl");

        log.debug("runTest:method=" + method + " local_file_path=" + local_file_path + " bucket=" + bucket + " object=" + object);

        SampleResult result = new SampleResult();
        result.sampleStart(); // start stopwatch

        try {
            ClientConfiguration config = new ClientConfiguration();
            if (proxy_host != null && !proxy_host.isEmpty()) {
                config.setProxyHost(proxy_host);
            }
            if (proxy_port != null && !proxy_port.isEmpty()) {
                config.setProxyPort(Integer.parseInt(proxy_port));
            }
            if (proxy_scheme != null && !proxy_scheme.isEmpty()) {
                if (proxy_scheme.toLowerCase().equals("http")) {
                    config.setProxyProtocol(Protocol.HTTP);
                } else if (proxy_scheme.toLowerCase().equals("https")) {
                    config.setProxyProtocol(Protocol.HTTPS);
                }
            }
            if (proxy_username != null && !proxy_username.isEmpty()) {
                config.setProxyUsername(proxy_username);
            }
            if (proxy_password != null && !proxy_password.isEmpty()) {
                config.setProxyPassword(proxy_password);
            }


            AmazonS3ClientBuilder builder = AmazonS3ClientBuilder.standard().withClientConfiguration(config).withRegion(region);
            if (key_id != null && !key_id.isEmpty()) {
                builder = builder.withCredentials(new StaticCredentialsProvider(new BasicAWSCredentials(key_id, secret_key)));
            }

            if (endpoint != null && !endpoint.isEmpty()) {
                builder = builder.withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(endpoint, region));
            }
            AmazonS3 s3Client = builder.build();
            ObjectMetadata meta = null;

            if (method.equals("GET")) {
                S3Object s3object = s3Client.getObject(bucket, object);
                S3ObjectInputStream stream = s3object.getObjectContent();
                if (local_file_path != null && !local_file_path.isEmpty()) {
                    Files.copy(stream, Paths.get(local_file_path), StandardCopyOption.REPLACE_EXISTING);
                } else {
                    if (object.endsWith(".avro")) {
                        try {
                            DatumReader<Object> reader = new GenericDatumReader<>();
                            DataFileStream<Object> streamReader = new DataFileStream<>(stream, reader);
                            Schema schema = streamReader.getSchema();
                            ByteArrayOutputStream output = new ByteArrayOutputStream();
                            DatumWriter<Object> writer = new GenericDatumWriter<>(schema);
                            JsonEncoder jsonEncoder = EncoderFactory.get().jsonEncoder(schema, output, true);

                            while (streamReader.hasNext()) {
                                Object datum;
                                datum = streamReader.next();
                                log.debug(datum.toString());
                                writer.write(datum, jsonEncoder);
                                jsonEncoder.flush();
                            }
                            output.flush();
                            log.info("Successfully parsed Avro to JSON");
                            result.setResponseData(output.toByteArray());
                        } catch(Exception e) {
                            log.warn("😔 Couldn't parse Avro " + e.getMessage());
                            result.setResponseData(StreamUtils.copyToString(stream, StandardCharsets.UTF_8), "UTF-8");
                        }
                    } else {
                        result.setResponseData(StreamUtils.copyToString(stream, StandardCharsets.UTF_8), "UTF-8");
                    }
                }
                stream.close();
                meta = s3object.getObjectMetadata();
            } else if (method.equals("PUT")) {
                File file = new File(local_file_path);
                PutObjectRequest por = new PutObjectRequest(bucket, object, file);
                if (acl != null && !acl.isEmpty()) {
                  por = por.withCannedAcl(CannedAccessControlList.valueOf(acl));
                }

                s3Client.putObject(por);
            }

            result.sampleEnd(); // stop stopwatch
            result.setSuccessful(true);
            if (meta != null) {
                result.setResponseMessage("OK on url:" + bucket + "/" + object + ". Length=" + meta.getContentLength() + ". Last Modified=" + meta.getLastModified() + ". Storage Class=" + meta.getStorageClass());
            } else {
                result.setResponseMessage("OK on url:" + bucket + "/" + object + ".No metadata");
            }
            result.setResponseCodeOK(); // 200 code

        } catch (Exception e) {
            result.sampleEnd(); // stop stopwatch
            result.setSuccessful(false);
            result.setResponseMessage("Exception: " + e);

            // get stack trace as a String to return as document data
            java.io.StringWriter stringWriter = new java.io.StringWriter();
            e.printStackTrace(new java.io.PrintWriter(stringWriter));
            result.setResponseData(stringWriter.toString(), "UTF-8");
            result.setDataType(org.apache.jmeter.samplers.SampleResult.TEXT);
            result.setResponseCode("500");
        }

        return result;
    }
}
